core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala

```diff
@@ -501,28 +530,63 @@ class SpendReportingService(
           }
           .toMap
       }
-
       _ = if (projectNames.isEmpty) {
         throw RawlsExceptionWithErrorReport(
           StatusCodes.NotFound,
           s"no spend data found for billing project ${project.value} between dates ${toISODateString(start)} and ${toISODateString(end)}"
         )
       }
 
-      query = getQuery(aggregations, spendExportConf)
-      queryJob = setUpQuery(query, spendExportConf, start, end, projectNames)
+      // is this spend report cacheable?
+      isCacheable = aggregationsAreCacheable(aggregations)
+
+      queryResults: Option[SpendReportingResults] <-
+        if (isCacheable) {
+          // look for a cached spend report
+          getCachedSpendReportData(projectNames, start, end) flatMap { cachedResults =>
+            if (cachedResults.isCacheValid) {
+              // cached spend report found; use it
+              rawlsCacheHitRate().hit()
+              Future.successful(cachedResults.results)
+            } else {
+              // cached spend report not found; go to BigQuery for results
+              rawlsCacheHitRate().miss()
+              getSpendForGCPBillingProjectFromBigQuery(aggregations,
+                                                       spendExportConf,
+                                                       start,
+                                                       end,
+                                                       projectNames,
+                                                       childContext
+              ) map { bqResults =>
+                // save the BigQuery results back to cache and return
+                // the writeback to cache is fire-and-forget
+                insertRecordsWithMissingSpendData(bqResults, projectNames, start, end)
+                bqResults
```

Comment:
Doug, this is where you were suggesting we wait for the cache to populate, then re-query the cache. I see why that would be consistent with the cache-hit case … but since we already have a case of returning results directly from BQ (when the query is not cacheable), I still felt this was ok.